src/eckit/stats/util/Mutex.h

```diff
@@ -0,0 +1,58 @@
+/*
+ * (C) Copyright 1996- ECMWF.
+ *
+ * This software is licensed under the terms of the Apache Licence Version 2.0
+ * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * In applying this licence, ECMWF does not waive the privileges and immunities
+ * granted to it by virtue of its status as an intergovernmental organisation nor
+ * does it submit to any jurisdiction.
+ */
+
+
+#pragma once
+
+#define ECKIT_GEO_ECKIT_THREADS
+
+#if defined(ECKIT_GEO_ECKIT_THREADS)
+#include "eckit/thread/AutoLock.h"
+#include "eckit/thread/Mutex.h"
+#else
+#include <mutex>
+#endif
+
+
+namespace eckit::stats::util {
+
+
+#if defined(ECKIT_GEO_ECKIT_THREADS)
```

Comment:
This is a very odd, and somewhat worrying block of code. Why do we need to do special stuff. Why us std::mutex insufficient. And what is the logic here.

If it really is necessary, this needs to be heavily documented as to why (in comments).

Reply:
This is due to the current threading model supprted in eckit (and the rest of our stacks), to be it's own implementation based on eckit::Mutex. This provides a #define that can change the threading model to stl, but for the stack we should do this in one go -- to me this is a separate discussion. It doesn't work, mixing the models. I've raised this issue over the last few years, several times, but gave up...